File: ios/SQCoreMotion.m
#import <React/RCTBridgeModule.h>

@interface RCT_EXTERN_MODULE (SQCoreMotion, NSObject)

RCT_EXTERN_METHOD(subscribeStepCount:
                  (RCTPromiseResolveBlock)resolve
                  reject:(RCTPromiseRejectBlock)reject)

RCT_EXTERN_METHOD(subscribeDistance:
                  (RCTPromiseResolveBlock)resolve
                  reject:(RCTPromiseRejectBlock)reject)

RCT_EXTERN_METHOD(unSubscribeStepCount:
                  (RCTPromiseResolveBlock)resolve
                  reject:(RCTPromiseRejectBlock)reject)

RCT_EXTERN_METHOD(unSubscribeDistance:
                  (RCTPromiseResolveBlock)resolve
                  reject:(RCTPromiseRejectBlock)reject)

@end

File: ios/SqCoreMotion.swift
import Foundation
import React

@objc(SQCoreMotion)
class SQCoreMotion: RCTEventEmitter {

    public static var emitter: RCTEventEmitter?
    private var coreMotionManager = SQCoreMotionManager()

    override init() {
        super.init()

        RNHealthTracker.emitter = self
    }

    @objc
    override static func requiresMainQueueSetup() -> Bool {
        return false
    }

    @objc
    override func supportedEvents() -> [String]! {
      return [
        "update_step_count",
        "update_distance",
        "update_activity_type"
      ]
    }

    @objc public func subscribeStepCount(
            resolve: @escaping RCTPromiseResolveBlock,
            reject: @escaping RCTPromiseRejectBlock
    ) {
        self.coreMotionManager
            .subscribeStepCount { stepCount in
                RNHealthTracker.emitter?.sendEvent(
                    withName: "update_step_count",
                    body: [stepCount]
                )
            } activityTypeHandler: { activityType in
                RNHealthTracker.emitter?.sendEvent(
                    withName: "update_activity_type",
                    body: [activityType.rawValue]
                )
            }
        resolve(true)
    }
}
